README
Authen-CAS-External version 0.06
================================

INSTALLATION
------------

To install this module, run the following commands:

    perl Build.PL
    ./Build
    ./Build test
    ./Build install

DESCRIPTION
-----------

Authen::CAS::External is an implementation of a client for use with
web sites that use JA-SIG's Central Authentication Service.

More information about the CAS protocol can be found at the
[JA-SIG CAS homepage](http://www.jasig.org/cas).

SUPPORT AND DOCUMENTATION
-------------------------

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Authen::CAS::External

You can also look for information at:

*   [RT, CPAN's request tracker](http://rt.cpan.org/NoAuth/Bugs.html?Dist=Authen-CAS-External)

*   [AnnoCPAN, Annotated CPAN documentation](http://annocpan.org/dist/Authen-CAS-External)

*   [CPAN Ratings](http://cpanratings.perl.org/d/Authen-CAS-External)

*   [Search CPAN](http://search.cpan.org/dist/Authen-CAS-External/)


COPYRIGHT AND LICENCE
---------------------

Copyright 2009 Douglas Christopher Wilson.

This program is free software; you can redistribute it and/or
modify it under the terms of either:

*   the GNU General Public License as published by the Free
    Software Foundation; either version 1, or (at your option) any
    later version, or

*   the Artistic License version 2.0.



Changes
Revision history for perl distribution Authen-CAS-External

0.07 2011-10-17
 [ENHANCEMENTS]
 - Added support to type TicketGrantingCookie for the format shown in
   http://www.ja-sig.org/wiki/display/CASUM/Clustering+CAS.
 - Support authenticating with user agents that do not have a cookie jar.

 [BUG FIXES]
 - Fixed bug where when Authen::CAS::External::UserAgent was destroyed the
   user agent that was attached to it would have now-useless handles still
   attached.

0.06 2010-04-08
 [BUG FIXES]
 - Fixed bug where if the CAS deployment was in a sub path the module would
   always fail to retrieve the response.

0.05 2010-01-20
 [ENHANCEMENTS]
 - Added support for notification scraping from the CAS website. This is
   Authen::CAS::External::Response->notification.
 - The cas_url attribute on Authen::CAS::External can now take a URL as a
   string and will automatically change it to a URI object.

0.04 2009-08-23
 [BUG FIXES]
 - When the user_agent attribute is modified, the handlers are properly
   removed from the previous user agent.

 [OTHER]
 - Added a FAQ document.
 - Changed required Moose to 0.89.

0.03 2009-08-12
 [ENHANCEMENTS]
 - Support LDAP Password Policy Enforcement add on on CAS servers from
   http://www.ja-sig.org/wiki/display/CASUM/LDAP+Password+Policy+Enforcement

 [OTHER]
 - Added more documentation of features (though they are not new).
 - Changed the license to be more clear.
 - New dependency of HTML::LinkExtract version 0.03.

0.02 2009-07-31
 [ENHANCEMENTS]
 - Service URL can now end in / and in /login and will work fine.

 [OTHER]
 - Removed dependency on version.pm.

0.01 2009-06-14
 - First release.


lib/Authen/CAS/External.pm
package Authen::CAS::External;

use 5.008001;
use strict;
use utf8;
use warnings 'all';

# Module metadata
our $AUTHORITY = 'cpan:DOUGDUDE';
our $VERSION   = '0.07';

use Authen::CAS::External::Library qw(TicketGrantingCookie);
use Moose 0.89;
use MooseX::StrictConstructor 0.08;
use MooseX::Types::Moose qw(Str);
use URI 1.22;

# Clean the imports are the end of scope
use namespace::clean 0.04 -except => [qw(meta)];

# Role

with 'Authen::CAS::External::UserAgent';

# Attributes

has password => (
	is  => 'rw',
	isa => Str,

	clearer   => 'clear_password',
	predicate => 'has_password',
	trigger   => sub { shift->clear_ticket_granting_cookie },
);
has ticket_granting_cookie => (
	is  => 'rw',
	isa => TicketGrantingCookie,

	clearer       => 'clear_ticket_granting_cookie',
	documentation => q{The Ticket Granting Cookie for the CAS user session},
	predicate     => 'has_ticket_granting_cookie',
);
has username => (
	is  => 'rw',
	isa => Str,

	clearer   => 'clear_username',
	predicate => 'has_username',
	trigger   => sub { shift->clear_ticket_granting_cookie },
);

# Methods

sub authenticate {
	my ($self, %args) = @_;

	# Splice out the variables
	my ($service, $gateway, $renew) = @args{qw(service gateway renew)};

	# Get the URI to request
	my $url = $self->service_request_url(
		(defined $gateway ? (gateway => $gateway) : () ),
		(defined $renew   ? (renew   => $renew  ) : () ),
		(defined $service ? (service => $service) : () ),
	);

	# Do not redirect back to service
	my $redirect_back = $self->redirect_back;
	$self->redirect_back(0);

	# Get the service
	my $response = $self->get($url);

	# Restore previous value
	$self->redirect_back($redirect_back);

	if (!$self->has_previous_response) {
		confess 'Failed retrieving response';
	}

	# Set our ticket granting ticket if we have one
	if ($self->previous_response->has_ticket_granting_cookie) {
		$self->ticket_granting_cookie($self->previous_response->ticket_granting_cookie);
	}

	# Return the last response
	return $self->previous_response;
}

sub get_cas_credentials {
	my ($self, $service) = @_;

	# This default callback stub simply returns the stored
	# credentials
	if (!$self->has_username) {
		confess 'Unable to authenticate because no username was provided';
	}

	if (!$self->has_password) {
		confess 'Unable to authenticate because no password was provided';
	}

	# Return username, password
	return $self->username, $self->password;
}

sub get_cas_ticket_granting_cookie {
	my ($self, %args) = @_;

	# Splice out the variables
	my ($username, $service) = @args{qw(username service)};

	# This default callback stub simply returns the stored
	# credentials
	if (!$self->has_ticket_granting_cookie) {
		return;
	}

	# Return ticket granting ticket
	return $self->ticket_granting_cookie;
}

# Make immutable
__PACKAGE__->meta->make_immutable;

1;

__END__

=head1 NAME

Authen::CAS::External - Authenticate with CAS servers as a browser
would.

=head1 VERSION

This documentation refers to version 0.07.

=head1 SYNOPSIS

  my $cas_auth = Authen::CAS::External->new(
      cas_url => 'https://cas.mydomain.com/',
  );

  # Set the username and password
  $cas_auth->username('joe_smith');
  $cas_auth->password('hAkaT5eR');

  my $response = $cas_auth->authenticate();

  my $secured_page = $ua->get($response->destination);

=head1 DESCRIPTION

Provides a way to authenticate with a CAS server just as a browser
would. This is useful with web scrapers needing to login to a CAS
site.

=head1 CONSTRUCTOR

This is fully object-oriented, and as such before any method can be used, the
constructor needs to be called to create an object to work with.

=head2 new

This will construct a new object.

=over

=item new(%attributes)

C<%attributes> is a HASH where the keys are attributes (specified in the
L</ATTRIBUTES> section).

=item new($attributes)

C<$attributes> is a HASHREF where the keys are attributes (specified in the
L</ATTRIBUTES> section).

=back

=head1 ATTRIBUTES

=head2 cas_url

This is the URL of the CAS site excluding /login. This can be a URI object
or a string of a URL.

=head2 password

This is the password to use for logging in to the CAS site. When set, this
clears the L</ticket_granting_cookie>.

=head2 ticket_granting_cookie

This is the ticket granting cookie to use for logging into the CAS site. This
can be set to log in with just the cookie and no username or password.

=head2 username

This is the username to use for logging in to the CAS site. When set, this
clears the L</ticket_granting_cookie>.

=head1 METHODS

=head2 authenticate

This method will authenticate against the CAS service using the already supplied
username and password and will return a
L<Authen::CAS::External::Response|Authen::CAS::External::Response> object.

This method takes a HASH with the following keys:

=over

=item gateway

This is a Boolean of if the gateway parameter should be sent to the CAS server.
The default is to not send any gateway parameter.

=item renew

This is a Boolean of if the renew parameter should be sent to the CAS server.
The default is to not send any renew parameter.

=item service

This is a string that specifies the service value to send to the CAS server.
The default is to not send any service parameter.

=back

=head2 get_cas_credentials

This method is not actually used, but is required for classes to consume the
L<Authen::CAS::External::UserAgent|Authen::CAS::External::UserAgent> role as
this class does. This method will return the currently set username and
password to the user agent.

=head2 get_cas_ticket_granting_cookie

This method is not actually used but is required for classes to consume the
L<Authen::CAS::External::UserAgent|Authen::CAS::External::UserAgent> role as
this class does. This method will return the currently set ticket granting
cookie if the username requested matches the username set (and always should).

=head1 DEPENDENCIES

=over 4

=item * L<Moose|Moose> 0.89

=item * L<MooseX::StrictConstructor|MooseX::StrictConstructor> 0.08

=item * L<MooseX::Types::Moose|MooseX::Types::Moose>

=item * L<URI|URI> 1.22

=item * L<namespace::clean|namespace::clean> 0.04

=back

=head1 AUTHOR

Douglas Christopher Wilson, C<< <doug at somethingdoug.com> >>

=head1 BUGS AND LIMITATIONS

Please report any bugs or feature requests to
C<bug-authen-cas-external at rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Authen-CAS-External>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

I highly encourage the submission of bugs and enhancements to my modules.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

  perldoc Authen::CAS::External

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Authen-CAS-External>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Authen-CAS-External>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Authen-CAS-External>

=item * Search CPAN

L<http://search.cpan.org/dist/Authen-CAS-External/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright 2009 Douglas Christopher Wilson.

This program is free software; you can redistribute it and/or
modify it under the terms of either:

=over 4

=item * the GNU General Public License as published by the Free
Software Foundation; either version 1, or (at your option) any
later version, or

=item * the Artistic License version 2.0.

=back


Makefile.PL
use inc::Module::Install;

name     'Authen-CAS-External';
all_from 'lib/Authen/CAS/External.pm';
author   'Douglas Christopher Wilson <doug@somethingdoug.com>';
license  'perl';

# Requirements for the tests in t/
build_requires 'Test::More';

# Requirements for the modules in lib/
requires
	'MooseX::Types::URI' => 0;

requires 'HTML::Form'                => '5.817';
requires 'HTML::TokeParser'          => '3.00';
requires 'HTTP::Status'              => '5.817';
requires 'LWP::UserAgent'            => '5.819';
requires 'Moose'                     => '0.89';
requires 'Moose::Role'               => '0.89';
requires 'MooseX::StrictConstructor' => '0.08';
requires 'MooseX::Types'             => '0.08';
requires 'MooseX::Types::Moose'      => 0;
requires 'Scalar::Util'              => '1.14';
requires 'URI'                       => '1.22';
requires 'URI::QueryParam'           => 0;
requires 'namespace::clean'          => '0.04';

# Requirements for the tests in xt/
author_requires
	'Test::More'     => 0,
	'Test::Requires' => 0.02;

# Enable tests to be in multi-level directories
tests_recursive;

WriteAll;

